### PR TITLE
Document SmartCache C++ example code

### PR DESCRIPTION
**Pull request by bpkroth**
It is currently quite difficult to follow what's going on across three different languages.

This PR attempts to heavily comment the SmartCache C++ example to help inform new users how the flow works across each component.

It also links some of the other component markdown documentation into the documentation linking structure for easier browsing on the web page.

### CODE REVIEW

**Review by grlap — APPROVED** [review]

File: source/Examples/SmartCache/SmartCache.SettingsRegistry/AssemblyInitializer.cs [hunk 1]
```diff
@@ -78,19 +135,25 @@ static AssemblyInitializer()
             optimizationProblem.Objectives.Add(
                 new OptimizationObjective
                 {
+                    // Tell the optimizer that we want to maximize hit rate.
+                    //
                     Name = "HitRate",
-                    Minimize = true,
+                    Minimize = false,
```
Comment (bpkroth) [inline source/Examples/SmartCache/SmartCache.SettingsRegistry/AssemblyInitializer.cs:141]: @byte-sculptor for visibility

File: documentation/Troubleshooting.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+### VSCode in WSL
+
+#### "Missing .Net SDK" message when executing `code .` in WSL
+
+The [Omnisharp](https://github.com/OmniSharp/omnisharp-vscode/wiki/Troubleshooting:-'The-.NET-Core-SDK-cannot-be-located.'-errors) plugin for VSCode may have trouble finding the `dotnet` setup locally for the MLOS repo in `tools/`, even if you source the `scripts/init.linux.sh` script to setup your `PATH` environment.
+
+To workaround this issue, you can [install `dotnet` system wide for your WSL2 distro](https://docs.microsoft.com/en-us/dotnet/core/install/linux).
+
+Here are the instructions for Ubuntu 20.04:\
+<https://docs.microsoft.com/en-us/dotnet/core/install/linux-ubuntu#2004->
```
Comment (bpkroth) [inline documentation/Troubleshooting.md:24]: Documenting some setup issues amueller ran into (#98).  Could get moved to its own PR.

File: source/Examples/SmartCache/Main.cpp [hunk 3]
```diff
@@ -6,27 +6,37 @@
...
+//
 #include "MlosPlatform.Std.inl"
 
 using namespace Mlos::Core;
 using namespace SmartCache;
 
-// Include codegen files.
-//
-#include "GlobalDispatchTable.h"
-
```
Comment (bpkroth) [inline source/Examples/SmartCache/Main.cpp:26]: This was already included in `stdafx.h`

File: source/Examples/SmartCache/SmartCache.SettingsRegistry/AssemblyInitializer.cs [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Security;
```
Comment (bpkroth) [inline source/Examples/SmartCache/SmartCache.SettingsRegistry/AssemblyInitializer.cs:11]: Not sure why this got added ...

File: source/Examples/SmartCache/SmartCache.SettingsRegistry/Codegen/SmartCache.cs [hunk 5]
```diff
@@ -5,6 +5,32 @@
...
+// optionally namespaces - see SmartCache/CodeGen/AssemblyInfo.cs) to allow
+// multiple settings registries to be included in the same project.
+//
+// Mlos.SettingsSystem.CodeGen outputs for multiple languages (e.g. C++, C#,
+// etc.)
+//
+// For C++ the codegen are expected to be used with #include statements.
+// For C# the codegen is combined with the code in this file to create a
+// loadable dll that the (C#) Mlos.Agent can directly use to process messages
+// and access settings in shared memory.
```
Comment (bpkroth) [inline source/Examples/SmartCache/SmartCache.SettingsRegistry/Codegen/SmartCache.cs:33]: Should probably also reference a `.md` file that has the full details on this.  It's really just meant as guideposts for newcomers first reading through the example code to hopefully help them jump through all the different languages.

File: source/Examples/SmartCache/stdafx.h [hunk 6]
```diff
@@ -6,15 +6,19 @@
 // @File: stdafx.h
 //
 // Purpose:
-//      <description>
+//  This is the standard include header, expected to be in every compilation unit
+//  in this project.
```
Comment (bpkroth) [inline source/Examples/SmartCache/stdafx.h:10]: This is not as common for non-Windows devs, so it seemed usful to explain a tad.

File: source/Mlos.Core/MlosContext.h [hunk 7]
```diff
@@ -126,7 +133,7 @@ class InternalMlosContextInitializer
 
     InternalMlosContextInitializer& operator=(const InternalMlosContextInitializer&) = delete;
 
-public:
+private:
```
Comment (bpkroth) [inline source/Mlos.Core/MlosContext.h:136]: @grlap for visibility

Something I noticed while reading through.  Seemed inappropriate for these fields to be public, but it's probably also inappropriate to include this change in this PR.

File: source/Mlos.SettingsSystem.CodeGen/CodeWriters/CppCodeWriter.cs [hunk 8]
```diff
@@ -46,12 +46,20 @@ public override void WriteComments(CodeComment codeComment)
         {
             WriteLine();
 
-            foreach (string lineComment in codeComment.Summary.Split(new[] { Environment.NewLine }, StringSplitOptions.None))
```
Comment (bpkroth) [inline source/Mlos.SettingsSystem.CodeGen/CodeWriters/CppCodeWriter.cs:49]: There's a missing `Summary != null` check here.
The change also adds Remarks (if available) to the codegen output.

File: .vscode/settings.json [hunk 9]
```diff
@@ -3,5 +3,10 @@
     "launch": {
         "configurations": [],
         "compounds": []
-    }
+    },
+    "C_Cpp.default.configurationProvider": "ms-vscode.cmake-tools",
+    "cmake.generator": "Unix Makefiles",
+    "cmake.preferredGenerators": [
+        "Unix Makefiles"
+    ]
```
Comment (bpkroth) [inline .vscode/settings.json:11]: Make vscode be consistent with our (and the default) cmake makefile generation mechanism.
Also probably doesn't belong here.

File: source/Examples/SmartCache/GlobalDispatchTable.h [hunk 10]
```diff
@@ -28,6 +31,18 @@ constexpr uint32_t SmartCache::ObjectDeserializationHandler::DispatchTableBaseIn
     return static_cast<uint32_t>(Mlos::Core::ObjectDeserializationHandler::DispatchTableElementCount());
 }
 
+// Note: As additional settings registries are included in the same project the
+// DispatchTableBaseIndex()s should be added together for each prior component
+// corresponding to the DispatchTables that are concatenated below.
+// See SmartSharedChannel/GlobalDispatchTable.h for an example.
+
+// Registers each of the code generated messages for the channel message handler
```
Comment (amueller) [inline source/Examples/SmartCache/GlobalDispatchTable.h]: Maybe this sentence should go first? My first question looking at this file is "what is a dispatch table"? Adding some more context might also help.

File: source/Examples/SmartCache/Main.cpp [hunk 11]
```diff
@@ -6,27 +6,37 @@
...
 //      <special-instructions>
 //
 //*********************************************************************
 
+// Include all the common headers for the application
+// (including Mlos.Core and component settings registry code generation output)
+//
 #include "stdafx.h"
 
+// Include platform specific implementations of some Mlos functions.
```
Comment (amueller) [inline source/Examples/SmartCache/Main.cpp:27]:
```suggestion
// Include platform specific implementations of some MLOS functions.
```

File: source/Examples/SmartCache/Main.cpp [hunk 12]
```diff
@@ -6,27 +6,37 @@
...
 using namespace SmartCache;
 
-// Include codegen files.
-//
-#include "GlobalDispatchTable.h"
-
 #include "SmartCacheImpl.h"
 #include "Workloads.h"
 
+// A basic Windows result code handler.
```
Comment (amueller) [inline source/Examples/SmartCache/Main.cpp:38]: The curious reader might ask: aren't we running this on linux? ;)

File: source/Examples/SmartCache/Main.cpp [hunk 13]
```diff
@@ -56,40 +76,82 @@ main(
     //
     ISharedChannel& feedbackChannel = mlosContext.FeedbackChannel();
 
+    // This background thread uses a lambda to monitor the feedback channel for
+    // new messages and process them using the callbacks registered for each
+    // message type in the global dispatch table.
+    //
     std::future<bool> feedbackChannelReader = std::async(
```
Comment (amueller) [inline source/Examples/SmartCache/Main.cpp:83]: how is the return value used? just do indicate that the thread finished?

File: source/Examples/SmartCache/Main.cpp [hunk 14]
```diff
@@ -56,40 +76,82 @@ main(
...
+    // config for this component and if not creates it.
+    //
     hr = mlosContext.RegisterComponentConfig(config);
     CheckHR(hr);
 
-    // Create an intelligent component.
+    // Create an instance of our SmartCache component to tune.
+    //
+    // Note that we pass it a ComponentConfig instance, which also includes our
+    // MlosContext instance, so that the component can internally send telemtry
```
Comment (amueller) [inline source/Examples/SmartCache/Main.cpp:147]:
```suggestion
    // MlosContext instance, so that the component can internally send telemetry
```

File: source/Examples/SmartCache/Main.cpp [hunk 15]
```diff
@@ -99,37 +161,72 @@ main(
...
+        // Note: the message (as defined in
+        // SmartCache.SettingsRegistry/CodeGen/SmartCache.cs) has no members, so
+        // there's no details to fill in here (and it is just zero-initialized).
+        // It's simply a signal to send to the external agent to request a new
+        // config be populated in the shared memory region.
         //
         SmartCache::RequestNewConfigurationMessage msg = { 0 };
         mlosContext.SendTelemetryMessage(msg);
 
+        // Now, we wait for the external agent to respond to our request.
```
Comment (amueller) [inline source/Examples/SmartCache/Main.cpp:211]: If no optimizer is registered then... ? it'll send a message without touching the shared memory?

File: source/Examples/SmartCache/SmartCache.SettingsRegistry/AssemblyInitializer.cs [hunk 16]
```diff
@@ -134,18 +207,19 @@ private static void RequestNewConfigurationMessageHandler(SmartCacheProxy.Reques
 
                     string currentConfigJsonString = JsonSerializer.Serialize(currentConfigDictionary, JsonOptions);
 
-                    // Register an observation.
+                    // Send (register) the observation with the optimizer.
```
Comment (amueller) [inline source/Examples/SmartCache/SmartCache.SettingsRegistry/AssemblyInitializer.cs:210]: "send with the optimizer" doesn't really work? with/to?

File: source/Examples/SmartCache/SmartCache.SettingsRegistry/Codegen/SmartCache.cs [hunk 17]
```diff
@@ -5,6 +5,32 @@
...
+// sent to out/Mlos.CodeGen.out/SmartCache/SettingsProvider_gen_*.*
+//
+// Different settings registries are separated to different folders (and
+// optionally namespaces - see SmartCache/CodeGen/AssemblyInfo.cs) to allow
+// multiple settings registries to be included in the same project.
+//
+// Mlos.SettingsSystem.CodeGen outputs for multiple languages (e.g. C++, C#,
+// etc.)
+//
+// For C++ the codegen are expected to be used with #include statements.
```
Comment (amueller) [inline source/Examples/SmartCache/SmartCache.SettingsRegistry/Codegen/SmartCache.cs:30]:
```suggestion
// For C++ the codegen is expected to be used with #include statements.

```
?

File: source/Mlos.Agent.Server/MlosAgentServer.cs [hunk 18]
```diff
@@ -75,15 +87,24 @@ public static void Main(string[] args)
...
             }
 
-            // Create circular buffer shared memory before running the target process.
+            // Create (or open) the circular buffer shared memory before running the target process.
             //
             MainAgent.InitializeSharedChannel();
 
-            // Active learning, almost done. In active learning the MlosAgentServer controls the workload against the target component.
+            // Active learning mode.
+            // In active learning the MlosAgentServer controls the workload against the target component.
```
Comment (amueller) [inline source/Mlos.Agent.Server/MlosAgentServer.cs:107]: I don't see how/where the workload is controlled. It just launches an executable, right?

File: source/Mlos.Agent.Server/MlosAgentServer.cs [hunk 19]
```diff
@@ -100,6 +121,24 @@ public static void Main(string[] args)
 
             Task grpcServerTask = CreateHostBuilder(Array.Empty<string>()).Build().RunAsync(cancellationTokenSource.Token);
 
+            // Start the MainAgent message processing loop as a background thread.
+            //
+            // In MainAgent.RunAgent we loop on the shared memory control and
+            // telemetry channels looking for messages and dispatching they to
```
Comment (amueller) [inline source/Mlos.Agent.Server/MlosAgentServer.cs:127]:
```suggestion
            // telemetry channels looking for messages and dispatching them to

```

File: source/Mlos.Agent.Server/MlosAgentServer.cs [hunk 20]
```diff
@@ -100,6 +121,24 @@ public static void Main(string[] args)
 
             Task grpcServerTask = CreateHostBuilder(Array.Empty<string>()).Build().RunAsync(cancellationTokenSource.Token);
 
+            // Start the MainAgent message processing loop as a background thread.
+            //
+            // In MainAgent.RunAgent we loop on the shared memory control and
+            // telemetry channels looking for messages and dispatching they to
+            // their registered callback handlers.
+            //
+            // The set of recognized messages are dynamically registered using
```
Comment (amueller) [inline source/Mlos.Agent.Server/MlosAgentServer.cs:130]:
```suggestion
            // The set of recognized messages is dynamically registered using

```

File: source/Mlos.Agent.Server/MlosAgentServer.cs [hunk 21]
```diff
@@ -100,6 +121,24 @@ public static void Main(string[] args)
...
+            //
+            // In MainAgent.RunAgent we loop on the shared memory control and
+            // telemetry channels looking for messages and dispatching they to
+            // their registered callback handlers.
+            //
+            // The set of recognized messages are dynamically registered using
+            // the RegisterSettingsAssembly method which is called through the
+            // handler for the RegisterAssemblyRequestMessage.
+            //
+            // Once registered, the SettingsAssemblyManager uses reflect to
```
Comment (amueller) [inline source/Mlos.Agent.Server/MlosAgentServer.cs:134]:
```suggestion
            // Once registered, the SettingsAssemblyManager uses reflection to

```

not sure?

**Review by amueller — COMMENTED** [review]
Looks good apart from minor nitpicks. I can't really comment on the C# changes.

File: documentation/Troubleshooting.md (hunk 2 above)
Comment (amueller) [reply documentation/Troubleshooting.md:24]: Could still not make this work after successfully running these instructions.

Comment (amueller) [reply documentation/Troubleshooting.md:24]: Ok actually works now :)

File: source/Mlos.Core/MlosContext.h (hunk 7 above)
Comment (bpkroth) [reply source/Mlos.Core/MlosContext.h]: Backed this out - it doesn't belong here.  Can take it up separately later on.

File: source/Examples/SmartCache/SmartCache.SettingsRegistry/AssemblyInitializer.cs (hunk 4 above)
Comment (bpkroth) [reply source/Examples/SmartCache/SmartCache.SettingsRegistry/AssemblyInitializer.cs]: Removed.

File: source/Examples/SmartCache/GlobalDispatchTable.h (hunk 10 above)
Comment (bpkroth) [reply source/Examples/SmartCache/GlobalDispatchTable.h]: I tried to rewrite all of the comments in this file.  Can you please take another look and see if it's any better?

File: source/Examples/SmartCache/Main.cpp (hunk 12 above)
Comment (bpkroth) [reply source/Examples/SmartCache/Main.cpp]: Tried to improve that.  Greg has a change coming that actually renames the function to something more meaningful as well.

File: source/Mlos.Agent.Server/MlosAgentServer.cs (hunk 18 above)
Comment (bpkroth) [reply source/Mlos.Agent.Server/MlosAgentServer.cs:107]: You're right, I think that is probably still future/planned work.  I'll tag it as such.

File: source/Examples/SmartCache/Main.cpp (hunk 13 above)
Comment (bpkroth) [reply source/Examples/SmartCache/Main.cpp:89]: It's not at the moment and ideally the compiler should complain about that (#22).

File: source/Examples/SmartCache/Main.cpp (hunk 15 above)
Comment (bpkroth) [reply source/Examples/SmartCache/Main.cpp:217]: There's a comment about that in `SmartCache.SettingsRegistry/AssemblyInitializer.cs`.

Basically, if there's no optimizer, the request is a no-op and just continues to use the existing value in shared memory.

Comment (bpkroth) [reply source/Examples/SmartCache/Main.cpp:217]: Do you think it needs mention of that here as well?

Comment (bpkroth) [reply source/Examples/SmartCache/Main.cpp:217]: added another comment to help them track the flow to the RequestNewConfigurationMessage handler in that AssemblyInitializer.cs file

File: website/build_site.sh [hunk 22]
```diff
@@ -74,17 +74,18 @@ cp ../LICENSE.txt content/
...
         source/ \
         test/ \
-        -name '*.md' \
-) | while read md_path; do
-    md_dir=$(dirname "$md_path")
-    md_file=$(basename "$md_path")
-    mkdir -p "content/$md_dir"
-    if [ "$md_file" == 'README.md' ]; then
+        -name '*.md' -or \
+        -name '*.svg' \
```
Comment (bpkroth) [inline website/build_site.sh:78]: Also copy the images over.

File: source/Examples/SmartCache/SmartCache.SettingsRegistry/AssemblyInitializer.cs (hunk 16 above)
Comment (bpkroth) [reply source/Examples/SmartCache/SmartCache.SettingsRegistry/AssemblyInitializer.cs]: reworked the wording

**Review by amueller — APPROVED** [review]
